File: bathysphere_numerical/simulate/physics/wind.py
from numpy import zeros, array


class Wind:
    def __init__(self, speed=0.0, delta=0.0, minimum=0.0):
        """
        Simulate wind speed and mixing.

        :param speed: starting wind speed
        :param delta: acceleration
        :param minimum: floor for clipping
        """
        self._speed = speed
        self._slope = delta
        self._min = minimum

    def _simple(self, speed=None):
        """
        Basic wind mixing rate.

        :param speed: optional override for internally tracked wind speed
        :return: mixing rate array or value
        """
        if speed is None:
            speed = self._speed

        return array(0.728 * speed ** 0.5 - 0.317 * speed + 0.0372 * speed ** 2)

    def _update(self, dt, clip=None):
        """
        Update velocity shear due to wind forcing, and optionally the clip values

        :return: success
        """
        self._speed += self._slope * dt

        if clip is not None:
            self._min = clip

        return True

    @staticmethod
    def _shear(velocity, topology, precision=float):
        """
        Calculate current velocity shear vector for selected cells

        :param velocity: velocity field, assumed to be already subset to surface and U, V components
        :param topology: tuple of parents of the node
        :param precision:

        :return:
        """

        n = len(topology)
        vectors = zeros((n, 2), dtype=precision)  # shear vectors

        for ii in range(n):
            parents = topology[ii]
            sq = velocity[parents, :, :].mean(axis=0) ** 2  # shape is (points, 3, layers, dim)
            vectors[ii] += sq.sum(axis=0) ** 0.5  # reduce to root of the sums, shape is (dim)

        return abs(vectors[:, 0] - vectors[:, 1])

    @classmethod
    def _dynamic(cls, nodes, layers, velocity, diffusivity=0.0):
        """

        :param nodes: object
        :param layers: object
        :param velocity: water velocity field
        :param diffusivity: of oxygen across air-water interface, m2/day

        :return: mixing rate
        """
        depth = nodes.depth * layers.z[:2].mean()
        subset = velocity[:, :2, :2]
        return (diffusivity * cls._shear(subset, nodes.parents)) / depth ** 0.5

    def mixing(self, dt=None, minimum=None, simple=True, kwargs=None, speed=None):
        """
        Update wind forcing if necessary. Determine mixing rate. Calculate and return aeration

        :param kwargs: depends on context
        :param dt: optional time step, triggers wind state update
        :param minimum: minimum value for mixing, m/day
        :param simple: use wind speed as proxy for mixing rate

        :return: aeration due to surface wind mixing
        """
        assert simple or kwargs is not None, "Cannot perform aeration calculation."

        if dt is not None:
            self._update(dt, clip=minimum)

        rate = self._simple(speed) if simple else self._dynamic(**kwargs)
        return rate.clip(min=self._min)

File: src/rust/wind.rs
pub mod wind_system {

    use crate::{Limit, Array};
    use crate::tessellate::tessellate::{Node, Layers};


    pub struct Wind {
        /*
        Simulate wind speed and mixing
        */
        speed: f64,
        delta: f64,
        limit: Limit,

    }

    impl Wind {

        fn new(max_speed: f64) -> Wind {
            Wind {
                speed: 0.0,
                delta: 0.0,
                limit: Limit::new(0.0, max_speed)
            }
        }

        fn update(&mut self, dt: f64) {
            self.speed += self.delta * dt;
        }

        fn simple_mixing(&self) -> f64 {
            /*
            Basic wind mixing rate.
            */
            0.728 * self.speed.powf(0.5) - 0.317 * self.speed + 0.0372 * self.speed.powi(2)
        }

        fn shear(velocity: Array, topology: Array) -> Array {

            /*
            Calculate current velocity shear vector for selected cells
        
            :param velocity: velocity field, assumed to be already subset to surface and U, V components
            :param topology: tuple of parents of the node
            :param precision:
            */
            
            let n = topology.len();
            let mut shear_vectors: Vec<[f64; 2]> = Vec::with_capacity(n);
    
            for parents in topology {
                
                let sq = velocity[parents, :, :].mean(0).powi(2);   // shape is (points, 3, layers, dim)
                shear_vectors[ii] += sq.sum(0) ** 0.5 // reduce to root of the sums, shape is (dim)
            }
    
            return (shear_vectors[:, 0] - shear_vectors[:, 1]).abs()
        
        }

        fn dynamic_mixing(&self, nodes: Vec<Node>, layers: Layers) -> Array {
           
            //     mapping: ((int), (int)),
            //     velocity: array,
            //     diffusivity: float or array = 0.0,
         
        
            //     :param layers: object
            //     :param velocity: water velocity field
            //     :param diffusivity: of oxygen across air-water interface, m2/day
        
            //     :return: mixing rate
            //     """

            nodes, layers = mapping
            depth = nodes.depth * &layers.z[..2].mean()
            subset = velocity[:, :2, :2]
            ((diffusivity * Wind.shear(subset, nodes.parents)) / depth ** 0.5).clip(
                min=self.limit.lower
            )
        }


    }


}